data_steward/retraction/retract_deactivated_pids.py

```diff
@@ -0,0 +1,441 @@
+        df_to_append = df_to_append.drop_duplicates()
+
+        # Create new df to loop through date time fields
+        df_to_iterate = pd.DataFrame(
+            columns=['project_id', 'dataset_id', 'table', 'column'])
+        df_to_iterate['project_id'] = date_fields_df['table_catalog']
+        df_to_iterate['dataset_id'] = date_fields_df['table_schema']
+        df_to_iterate['table'] = date_fields_df['table_name']
+        df_to_iterate['column'] = date_fields_df['column_name']
+
+        # Remove person table and death table
+        df_to_append = df_to_append[~df_to_append.table.str.
+                                    contains('death', 'person')]
+        df_to_iterate = df_to_iterate[~df_to_iterate.table.str.
+                                      contains('death', 'person')]
+
+        # Filter through date columns and append to the appropriate column
+        for i, row in df_to_iterate.iterrows():
+            column = getattr(row, 'column')
+            table = getattr(row, 'table')
+            if 'start_date' in column:
+                df_to_append.loc[df_to_append.table == table,
+                                 'start_date_column'] = column
+            elif 'end_date' in column:
+                df_to_append.loc[df_to_append.table == table,
+                                 'end_date_column'] = column
+            else:
+                df_to_append.loc[df_to_append.table == table,
+                                 'date_column'] = column
+
+        date_fields_info_df = date_fields_info_df.append(df_to_append)
+
+    return date_fields_info_df
+
+
+def is_deid_dataset(dataset_id):
+    """
+    Determine if a dataset is deid using regex defined above
+    :param dataset_id: dataset_id to verify if it is deid
+
+    :return: boolean
+    """
+    return bool(re.match(DEID_REGEX, dataset_id))
+
+
+def get_research_id(project, dataset, pid):
+    """
+    For deid datasets, this function queries the _deid_map table in the associated combined dataset based on the release
+    regex prefix; to return the research_id for the specific pid passed
+
+    :param project: bq name of project
+    :param dataset: bq name of dataset
+    :param pid: person_id passed to receive associated research_id
+    :return: research_id or None if it does not exist for that person_id
+    """
+    # Get non deid dataset prefix
+    prefix = dataset.split('_')[0]
+    prefix = prefix.replace('R', '')
+
+    research_id_df = bq.query(
+        RESEARCH_ID_QUERY.format(project=project, prefix_regex=prefix, pid=pid))
+    if research_id_df.empty:
+        logging.info('no research_id associated with person_id: %s ' % pid)
+        return None
+    else:
+        return research_id_df.to_string()
+
+
+def check_pid_exist(pid, date_row):
+    """
+    Queries the table that retraction will take place, to see if the PID exists before creating query
+    :param pid: person_id
+    :param date_row:
+    :return:
+    """
+    check_pid_df = bq.query(
```

Comment:
Check your pid before using it.  If the pid is either `None` or an empty string, `''`,  the formatted string will still be handled correctly by python but will cause a BigQuery failure.  If, for some reason, `get_research_id` returned `None` or an empty string both of these queries could be built by this function.  Both would fail.  You can check for this and return `False` without even querying BigQuery.  Or you can query BigQuery and catch the exception.  Then use the catch to set your return value.
```
SELECT
COUNT(*) AS count
FROM `aou-res-curation-test.lrwb_deid_test_input_20190910.observation`
WHERE person_id = None

BigQuery returns this error response:  google.api_core.exceptions.BadRequest: 400 Unrecognized name: None at [5:19]


SELECT
COUNT(*) AS count
FROM `aou-res-curation-test.lrwb_deid_test_input_20190910.observation`
WHERE person_id =

BigQuery returns this error response:  google.api_core.exceptions.BadRequest: 400 Syntax error: Unexpected end of script at [5:18]
```

Reply:
I have adjusted where if the `research_id` results in None it will not run `check_pid_exist` and exit the loop.